src/Microsoft.TemplateEngine.Cli/TemplateResolution/ListOrHelpTemplateListResolutionResult.cs

```diff
@@ -0,0 +1,241 @@
+// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using Microsoft.TemplateEngine.Edge.Template;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.TemplateEngine.Utils;
+using System.Collections.ObjectModel;
+using Microsoft.TemplateEngine.Abstractions;
+
+namespace Microsoft.TemplateEngine.Cli.TemplateResolution
+{
+    /// <summary>
+    /// Class represents the template list resolution result to be used for listing or displaying help purposes
+    /// </summary>
+    public class ListOrHelpTemplateListResolutionResult
+    {
+        public ListOrHelpTemplateListResolutionResult(string templateName, IReadOnlyCollection<ITemplateMatchInfo> coreMatchedTemplates)
+        {
+            _templateName = templateName;
+            _coreMatchedTemplates = coreMatchedTemplates;
+        }
+
+        private readonly string _templateName;
+        private readonly IReadOnlyCollection<ITemplateMatchInfo> _coreMatchedTemplates;
+        private IReadOnlyCollection<ITemplateMatchInfo> _exactMatchedTemplates;
+        private IReadOnlyCollection<ITemplateMatchInfo> _partiallyMatchedTemplates;
+
+        /// <summary>
+        /// Returns list of exact or partially matched templates by name and exact match by language, filter, baseline (if specified in command paramaters)
+        /// </summary>
+        public IReadOnlyCollection<ITemplateMatchInfo> ExactMatchedTemplates
+        {
+            get
+            {
+                if (_exactMatchedTemplates == null)
+                {
+                    _exactMatchedTemplates = GetCoreMatchedTemplatesWithDisposition(x => x.IsMatch);
```

Comment:
If this class is immutable, can we cache everything in constructor and just use that data without doing any more calculations?

Reply:
In most of the cases only exact matched templates will be calculated, however not sure how much performance can be saved on it.